upload/upload.go

```diff
@@ -4,86 +4,136 @@ import (
+		return "", "", err
 	}
 
 	return origUrl, thumbUrl, nil
 }
 
-func generateThumbnail(image image.Image, contentType string) ([]byte, error) {
-	thumbImage := imaging.Thumbnail(image, 500, 500, imaging.NearestNeighbor)
-
-	thumbBuf := new(bytes.Buffer)
-	err := imaging.Encode(thumbBuf, thumbImage, getImageFormat(contentType), imaging.JPEGQuality(80))
-	if err != nil {
-		return nil, fmt.Errorf("failed to encode thumbnail image: %w", err)
-	}
-
-	return thumbBuf.Bytes(), nil
-}
-
 // Delete removes the specified image files from the upload store.
-func Delete(driver Driver, recipeId int64, imageName string) error {
+func (u ImageUploader) Delete(recipeId int64, imageName string) error {
 	origPath := filepath.Join(getDirPathForImage(recipeId), imageName)
-	if err := driver.Delete(origPath); err != nil {
+	if err := u.Driver.Delete(origPath); err != nil {
 		return err
 	}
 	thumbPath := filepath.Join(getDirPathForThumbnail(recipeId), imageName)
-	return driver.Delete(thumbPath)
+	return u.Driver.Delete(thumbPath)
 }
 
 // DeleteAll removes all image files for the specified recipe from the upload store.
-func DeleteAll(driver Driver, recipeId int64) error {
+func (u ImageUploader) DeleteAll(recipeId int64) error {
 	dirPath := getDirPathForRecipe(recipeId)
-	err := driver.DeleteAll(dirPath)
+	err := u.Driver.DeleteAll(dirPath)
 
 	return err
 }
 
+// Load reads the image for the given recipe, returning the bytes of the file
+func (u ImageUploader) Load(recipeId int64, imageName string) ([]byte, error) {
+	origPath := filepath.Join(getDirPathForImage(recipeId), imageName)
+
+	file, err := u.Driver.Open(origPath)
+	if err != nil {
+		return nil, err
+	}
+
+	return ioutil.ReadAll(file)
+}
+
+func (u ImageUploader) generateThumbnail(original image.Image, contentType string, saveDir string, imageName string) (string, error) {
+	thumbImage := imaging.Thumbnail(original, u.imgCfg.ThumbnailSize, u.imgCfg.ThumbnailSize, toResampleFilter(u.imgCfg.ThumbnailQuality))
+
+	thumbBuf := new(bytes.Buffer)
+	err := imaging.Encode(thumbBuf, thumbImage, getImageFormat(contentType), imaging.JPEGQuality(toJPEGQuality(u.imgCfg.ThumbnailQuality)))
+	if err != nil {
+		return "", fmt.Errorf("failed to encode thumbnail image: %w", err)
+	}
+
+	return u.saveImage(thumbBuf.Bytes(), saveDir, imageName)
+}
+
+func (u ImageUploader) generateFitted(original image.Image, contentType string, saveDir string, imageName string) (string, error) {
+	var fittedImage image.Image
+
+	bounds := original.Bounds()
+	if bounds.Dx() <= u.imgCfg.ImageSize && bounds.Dy() <= u.imgCfg.ImageSize {
+		fittedImage = original
+	} else {
+		fittedImage = imaging.Fit(original, u.imgCfg.ImageSize, u.imgCfg.ImageSize, toResampleFilter(u.imgCfg.ImageQuality))
+	}
+
+	fittedBuf := new(bytes.Buffer)
+	err := imaging.Encode(fittedBuf, fittedImage, getImageFormat(contentType), imaging.JPEGQuality(toJPEGQuality(u.imgCfg.ImageQuality)))
```

Comment:
Consider always saving as JPEG, regardless of source format. That may in fact be required for this change.

Reply:
Considered. Will leave as-is for now, since it is consistent with previous code. Still might not be correct though.